### PR TITLE
Replace assert with exceptions - addresses #28

### PR DESCRIPTION
**Pull request by wxmerkt**
I took a first stab at addressing #28 since I missed some argument errors I made today. I am not privy to the ongoing discussions, but perhaps this is a first step at moving ahead? If not, please feel free to close.

### CODE REVIEW

File: include/curves/optimization/details.h [hunk 1]
```diff
@@ -174,8 +174,12 @@ problem_data<Point, Numeric, Safe> setup_control_points(const problem_definition
   // add remaining variables (only if no end_pos constraints)
   for (; i < numControlPoints; ++i) variables_.push_back(var_t::Zero(pDef.dim_));
 
-  assert(numControlPoints > numConstants);
-  assert(numControlPoints == variables_.size());
+  if (numControlPoints < numConstants) {
```
Comment (pFernbach) [inline include/curves/optimization/details.h]: Shouldn't it be <= ?

File: include/curves/optimization/details.h [hunk 2]
```diff
@@ -273,8 +281,9 @@ quadratic_variable<Numeric> bezier_product(In PointsBegin1, In PointsEnd1, In Po
   typedef Eigen::Matrix<Numeric, Eigen::Dynamic, 1> vector_x_t;
   unsigned int nPoints1 = (unsigned int)(std::distance(PointsBegin1, PointsEnd1)),
                nPoints2 = (unsigned int)(std::distance(PointsBegin2, PointsEnd2));
-  assert(nPoints1 > 0);
-  assert(nPoints2 > 0);
+  if (nPoints1 < 0 || nPoints2 < 0) {
```
Comment (pFernbach) [inline include/curves/optimization/details.h]: Same as above, should be <= . if ==0 the line above will create an error.

**Review by pFernbach — CHANGES_REQUESTED** [review]
Thanks for your work. I agree with the general direction of raising error instead of using assert for cases which will create obvious errors. 

Once the requested changes are made I will merge it.

**Comment by wxmerkt** [comment]
Thank you for the review. I have made the requested changes

**Review by pFernbach — APPROVED** [review]

**Pull request merged**